teps/0088-result-summaries.md

```diff
@@ -0,0 +1,392 @@
+
+  // Common Record agnostic fields.
+  google.protobuf.Timestamp start_time = 3;
+  google.protobuf.Timestamp end_time = 4;
+
+  enum Status {
+    UNKNOWN = 1;
+    SUCCESS = 2;
+    FAILURE = 3;
+    ...
+  }
+  Status status = 5;
+
+   // Key-value pairs representing arbitrary underlying record data that clients want to include
+   // that aren't covered by the above Record agnostic fields.
+   map<string, string> record_data = 6;
+}
+```
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this TEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a TEP within the wider Tekton community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+We want to let UIs/CLIs to show high level summaries of Results, e.g. -
+
+| Result | Type        | Status  | Duration |
+| ------ | ----------- | ------- | -------- |
+| A      | PipelineRun | SUCCESS | 30s      |
+| B      | TaskRun     | FAILURE | 10s      |
+| C      | CustomRun   | SUCCESS | 5s       |
+
+However, since Results do not currently store any Record information, the only
+mechanism to output such a UI would be to:
+
+1. List all Results
+2. For each Result, list all Records
+3. From the Records, figure out the order of precedence to know what to display
+   (e.g. if Record contains PipelineRun+TaskRun(s), show the PipelineRun)
+
+Instead, we want to provide fields to allow Result producers to give Result
+consumers consistent access to common high level data (i.e. status, timing, etc)
+and direction as to what Records are most important.
+
+### Goals
+
+<!--
+List the specific goals of the TEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+- Allow for Results to access high-level information (e.g. statuses, timing
+  information, primary Record data) in an efficient way.
+
+### Non-Goals
+
+<!--
+What is out of scope for this TEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Include all Records in a Result - we want to avoid:
+  1. making the size of a Result linear to the number of Records.
+  2. causing contention of the Result if there are many Records being updated at
+     once.
+- Include detailed Record information in a Result - we will continue to defer to
+  the Record to store complete information.
+
+### Open Questions
+
+- Should a Result be able to have multiple summaries? e.g. what if I have a
+  Result with multiple TaskRuns?
+  - How should a UI / tool handle this?
+- What fields should we include in a summary? Should things like Git references
+  that might not be applicable to all Results get special treatment, or should
+  they be relegated to `annotations`?
+- Do we need a mechanism to ensure that a Result stays in sync with the
+  underlying Record? What happens if they diverge?
```

Comment:
I'm not totally clear on this one - given the use case is supporting UIs/CLIs that report this information it seems like this is a necessary requirement?  In the event that we don't keep a Record Summary in-step with the underlying Record how would a UI / CLI detect this and know to look up the root Record and up-to-date info?

Reply:
Not necessarily. The idea is to provide some standardized fields that are common regardless of the underlying record type (with the exception of `annotations`, which is the catch-all generic/opaque map you describe).

One approach is to keep them separate and rely on clients to populate. This makes things like transforming values easier e.g. if we want to convert PipelineRun/TaskRun error reasons to a common error type.

The other option is to have a reference to fields within the Record, (e.g. similar to [additionalPrinterColumns](https://github.com/tektoncd/pipeline/blob/1ecb48215f4dd07bdf2567b5e7c8f7e4939da13b/config/300-pipelinerun.yaml#L73-L85), but this adds complexity to Result retrieval (e.g. for every Result fetch you need to also fetch the Result summary Record).